---
title: "Assignment 3"
author: "Group 28: Mona Behrouzian, Traye Lin, Maya Ansu"
date: "November 18, 2023"
output:
  pdf_document: default
  word_document: default
---


```{r setup, include=FALSE}
#Set up code
knitr::opts_chunk$set(echo = TRUE)

#code from TA to make sure my commented code does not fall off the PDF page
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=70), tidy=TRUE)

setwd("C:/Users/monab/Documents/R/Assignment3")

RNAseq <- read.csv("single_cell_RNA_seq_bam.sam", nrows=73, sep="\t", header=FALSE, fill=TRUE) 
#These lines contain tabulated information about the BAM file and the circumstances of its data collection.

head(RNAseq)
```

# Q1. Sequencing technologies
## Why are areas of the genome with high GC content are hard to sequence?
*If a genome has high GC content (i.e. high amount of guanine and cytosine base pairs) then the melting temperature is high as well (due to forming more secondary structures that are very stable). This means it is harder to separate the 2 strands, thus it is harder to do PCR denaturation. If the strands don't separate, they won't be amplified and read in NGS.*
*Additionally, high GC content in the entire genome influences fragment count, making these specific fragments underrepresented. Signal strength decreases as sequencing continues with high GC.*

*Source: Benjamini Y, Speed TP. Summarizing and correcting the GC content bias in high-throughput sequencing. Nucleic Acids Res. 2012 May;40(10):e72. doi: 10.1093/nar/gks001. Epub 2012 Feb 9. PMID: 22323520; PMCID: PMC3378858.*

# Q2. Global alignment exercise
## Similar to the approach for Needleman–Wunsch algorithm, find the best global alignment between the two following sequences: ATTCGAC and ATCAC. Use a gap penalty of -2 and the following scoring matrix
## In your answer, please include the grid table (used for storing the scores and traceback) and also include how you calculated the first top-left 9 elements of the table.
*see merged PDF on final pages*

# Q3. Looking at the Metadata of an alignment (SAM) file
## Q3.1. According to the header table in section 1.3 of the BAM/SAM document in the appendix, what do the SN and LN tags indicate?
*The SN tags indicate the reference sequence name. They all must be distinct.*
*The LN tags indicate the reference sequence length. It has a range of [1, (2^31) − 1]*

A sequence is any template string of bases to which we can align a read. This includes chromosomes (which are continuous sequences of bases) and new strings resulting from genetic modifications.
## Q3.2. What is the length of the X chromosome, in bp, for our alignment?
*The length of the X chromosome is 171031299*
```{r}
x_length <- RNAseq$V3[RNAseq$V2 == "SN:X"]
x_length
```


# Q4. Looking at the Reads of an alignment (SAM) file

## Q4.1. How many reads are there in this BAM file?
*There are 146,346 reads in this BAM file.*
```{r}
#load the reads into an R dataframe. Each row contains one read.
sam <- read.csv("single_cell_RNA_seq_bam.sam", sep="\t", header=FALSE, comment.char="@", col.names = paste0("V",seq_len(30)), fill=TRUE)
sam <- sam[paste0("V",seq_len(11))]

#dim(sam)
#head(sam)
print(nrow(sam))
```
 
## Q4.2.1 Which column of your dataframe should you look at to find the chromosome to which the read was aligned? 
*Column 3 shows the "RNAME" which would show us the chromosome to which the read was aligned.*

## 4.2.2 To which BAM data field does the dataframe column "V11" correspond?
*Column 11 corresponds to the "QUAL" tag, or the base quality, in ASCII format.*
```{r}
#Print out the 10th row of the dataframe to look at the format of a read.
sam[10,]
#Compare it to the mandatory BAM fields table in section 1.4 of the SAM/BAM documentation in the appendix. 
#The order of columns in the bam file have been preserved in the dataframe. 
```

## Q4.3. How many reads in this file align to chromosome X?
Hint: You can compare a column vector to a constant using logical symbols (==, <, >, etc.) to get a
column vector of TRUE or FALSE. Remember, when summing, a true symbol is worth "1" while a
false symbol is worth "0".
*There are 5,999 reads in this file that align to chromosome X.*
```{r}
Xchrom <- which(sam$V3 == 'X') #this is giving the indices of elements that satisfy the condition
print(length(Xchrom))
#head(Xchrom)
```
# Q4.4. What is the mean base quality (BQ) for reads aligning to chromosome X?
*32.7 is the mean BQ for reads aligning to chrom X.*
```{r}
Xchrom <- sam[sam$V3 == "X", ]

bq_int_x <- sapply(Xchrom$V11, utf8ToInt)-33

print(mean(bq_int_x))
```

## Q4.5. Plot the distribution of BQs across all bases and reads as a boxplot. 
Comment on your observation. 
Hint: This is similar to a boxplots that was provided in the lecture related to primary analysis. 
*We observe that there is a good per base sequence quality, throughout all the positions.*
*This shows us that we likely do not need to trim any of the data*
```{r}
bq_int <- sapply(sam$V11, utf8ToInt)-33
bq_int <- t(bq_int)

#dim(bq_int)

#We got help on our code from the TA! (Tina) :)
boxplot(bq_int[,1:10], at = 1:10, xlim = c(1,58), xaxt = 'n', main = "Distribution of Base Qualities", xlab = "Position in the read", ylab = "Base Quality")
boxplot(bq_int[,11:20], at = 11:20, add = TRUE, xaxt = 'n')
boxplot(bq_int[,21:30], at = 21:30, add = TRUE, xaxt = 'n')
boxplot(bq_int[,31:40], at = 31:40, add = TRUE, xaxt = 'n')
boxplot(bq_int[,41:50], at = 41:50, add = TRUE, xaxt = 'n')
boxplot(bq_int[,51:58], at = 51:58, add = TRUE, xaxt = 'n')

axis(1:58, at = 1:58)
```

## Q4.6. Referring to section 1.4 of the SAM/BAM documentation, what column contains the leftmost
mapping position of the reads?
*Column 4, with the POS tag.*

## Q4.7. In order to transform a BAM file into expression levels for each gene, we need to count the
number of reads covering a particular location or gene. The protein Hspa8 is located on
chromosome 9 at bases 40801273 - 40805199. How many reads have their leftmost mapping
position aligned within these coordinates?
Hint: you can implement AND logic on two column vectors with "&".
*119 reads.*
```{r}
reads_indices <- which(sam$V3 == '9' & sam$V4 >= 40801273 & sam$V4 <= 40805199)
reads_pos <- sam[reads_indices,4]

length(reads_pos)
```
Mapping quality is an indication of how well a read aligned to the reference genome during
the alignment step of processing our library data. It is reported as an integer between 0 and 255.
## Q4.8. How many reads have mapping quality less than 50? 
*61,527 reads.*
```{r}
mapq_less_50 <- which(sam$V5 < 50)
length(mapq_less_50)
```

## Q4.9. What is the mean mapping quality of the reads which have mapping quality less than 50?
Hint: you can obtain a subset of a dataframe by using df[bool_vec,] where bool_vec contains TRUE/FALSE elements and bool_vec and df have the same number of rows.
*0.2418125 is the mean mapping quality.*
```{r}
reads_mapq_less_50 <- sam$V5[mapq_less_50]
mean(reads_mapq_less_50)
```

## Q4.10. (bonus): The genome of the mouse used in this experiment has been edited to include the
DNA sequence for the protein 'tdTomato', which is a fluorophore. Count the number of reads
which align to the tdTomato sequence. Assuming that these reads are accurate, would you expect
this cell to emit fluorescently? What might be the purpose of modifying a genome to include a
fluorophore?
Hint: Think about studying cell populations under a microscope.
*asdf*
```{r}

```

# Q5. Investigating the Variants
We have used Strelka, which is a variant-calling tool, to find all of the SNPs and short indels in the
genome of this cell using the BAM file. The variants were then annotated using snpEff to label
them with information such as which gene they affect and the type of modification they result in
once the RNA undergoes translation to a protein. The results are in a VCF file (extension '.vcf')
which is attached. 

Use the following lines of code to obtain the header of the file and a dataframe where each
row is a variant. As you can see, information in the VCF file is organised by multiple levels of
character-separated data, so it will take multiple rounds of parsing to extract relevant
information.

```{r}
vcf_con <- file("RNA_seq_annotated_variants.vcf", open="r")
vcf_file <- readLines(vcf_con)
close(vcf_con)
vcf <- data.frame(vcf_file)
header <- vcf[grepl("##", vcf$vcf_file), ]
factor(header)
variants <- read.csv("RNA_seq_annotated_variants.vcf", skip=length(header),
header=TRUE, sep="\t")
```


## Q5.1 For the first variant (row) in the dataframe, what is the reference allele base at the site, and what is the alternative allele called by Strelka?
Hint: Take a look at the VCF Variant Call Format document in the appendix for details on each
column name.
*G and A*
```{r}
ref.allele <- variants$REF[1]
alt.allele <- variants$ALT[1]

cat("Reference Allele:", ref.allele, "\n")
cat("Alternative Allele:", alt.allele)
```

The INFO field is organised into variables by the form 'TAG=value' (see the VCF Variant Call Format document).
## Q5.2. Write code to obtain the entirety of the ANN info value contents from the INFO field for the first variant. 
```{r}
info_field <- as.character(variants$INFO[1])
info_variables <- strsplit(info_field, ";")[[1]]
ann_info <- info_variables[grep("ANN=", info_variables)]
ann_info
```

Q5.3. Each INFO tag-value pair is detailed in a line of the header, beginning with the tag '##INFO=
<ID=VARIABLE, ...'. Look for the header entry starting with '##INFO=<ID=ANN, ...' which details the
format of the ANN value contents. This tag-value pair contains the results of the annotations
found by snpEff. 
Based on the ANN value of the first variant, what does the 'Annotation' field tell us about this variant?
*The field tells us that this is an intronic variant.*
```{r}
#need to split up the annotation with commas
strsplit(as.character(ann_info), ",")
```

Hint: snpEff can return multiple annotation entries for the same variant because some variants
may have multiple possible effects. The first annotation entry is the most confident/important
and, resultantly, you should only look at the first entry to answer this and all subsequent
question. You can use strsplit() again with ',' separation character if you wish to look at each of the
ANN entries separately.
Hint: Refer to the snpEff documentation in the appendix for a list of snpEff annotation label
names and summaries of their effects.


Q5.4. Perform the parsing done in Q5.1-3 again on variant line 683. What gene would this variant affect?
*Rps19 gene. We can see it is this gene from the annotations printed below.* 
```{r}
ANN_683 = unlist(strsplit(as.character(variants$INFO)[683], "ANN="))[2]
strsplit(as.character(ANN_683), ",")
```


Q5.5. Within the entire VCF file, how many HIGH impact variants we see in total? 
*4 in total*
```{r}
high_impacts <- grep("HIGH", variants$INFO)

length(high_impacts)
```

Q5.6. What is a frameshift variant? Does it have a greater or lesser effect on the resultant protein than a missense variant? Why? 
*A frameshift variant is a type of mutation that involves the insertion or deletion of nucleotides that changes the reading frame of a DNA sequence. It often results in a premature STOP codon. A frameshift variant is more likely to have more severe consequences than the a missense variant. This is because a frameshift variant has the ability to affect the entire sequence downstream of the mutation, causing a shift in the entire amino sequence during protein translation. However, a missense variant involves the substitution of only one nucleotide, resulting in the change of a single amino acid in the protein sequence.*

Q5.7. 
We can divide variants into two broad categories: intronic/intergenic and exonic. 
Count the number of potential intronic variants AND intergenic! 
What do you notice about the number of intronic variants (compared to overall number of variants)?
Hint: Use grepl() on the INFO field to look for tell-tale tags.
Hint: assume no overlap between exonic and intronic tags within a variant entry.

*We observe that a majority of the variants are of the intronic/intergenic category.*
```{r}
intronic_variants <- grep("intron_variant", variants$INFO)
intergenic_variants <- grep("intergenic_region", variants$INFO)

cat("Number of intronic variants:", length(intronic_variants), "\n")
cat("Number of intergenic variants:",length(intergenic_variants), "\n")
total_interon_and_intergenic_variants <- length(intronic_variants) + length(intergenic_variants)
cat("Total number of intronic/intergenic:",total_interon_and_intergenic_variants, "\n")
cat("Total number of variants:", nrow(variants), "\n")
```

Q5.8. List all the genes that have been affected by coding mutations and have high impact. What do you find that is interesting? 
Hint: You can use SNPeff HIGH/MODERATE impact field to help you finding those genes.
*As printed below, the genes affected are: Nufip2, Tmem45a, Rpl11, n-R5s193-Spsb1.*
```{r}
coding_mutations <- grep("protein_coding", variants$INFO)
coding_mutations_and_high_impact <- grep("HIGH", variants$INFO[coding_mutations])

info_coding <- as.character(variants$INFO[coding_mutations_and_high_impact])
info_coding <- strsplit(info_coding, ";")

#printing off the annotation 
for (i in 1:length(info_coding)) {
  print(info_coding[[i]][grep("^ANN=", info_coding[[i]])])
}
```


Q5.9. (bonus): Using Strelka on our data, we can detect indels, but only to a limited extent. Most of the reads in our BAM file have read lengths around 60bp long. Why might this have consequences for the detection of insertions that are longer than 60bp?
*Due to the average read length being around 60bp long, it may be difficult to detect insertion longer than 60bp. This is because the variant caller may not have enough information to identify and characterize long insertions. Additionally, if the alignment algorithm relies on short reads it may be difficult to accurately align longer insertion to a reference genome.*

Q5.10. Variant Allele Frequency (VAF) is an important metric that helps us to measure how many
DNA molecules in a given sample are carrying a given variant. It also helps to identify potential
false-positive situations caused by incorrect base calls or alignment. VAF is calculated by
The number of variant alleles /(The number of Variant alleles + The number of Reference alleles)

In the form of a boxplot, plot the distribution of the VAFs across all the variants. How many
variants have VAF > 5%? How many of those variants (the ones with >5% VAF) are in coding
regions?
Hint: You will need to parse the genotype encoding field (GT:GQ:GQX:DP:DPF...) to get allele
counts and then get VAF. To understand that column, look at the VCF Variant Call Format
Document (GATK) section 5. 

More information on coding regions: https://en.wikipedia.org/wiki/Coding_regionLinks to an external site.
Coding mutations can be found from the table below. source: https://pcingola.github.io/SnpEff/se_inputoutput/Links to an external site.  

*How many variants have VAF > 5%? ----> 422 *
*How many of those variants (the ones with >5% VAF) are in coding regions? ----> 377*
```{r}
#Note, we got lots of help from the TA (Tina) :)
VAF <- c()

for (i in 1:nrow(variants)) {
  # split by : for format and ws20171223...
  # unlist both of those
  format_info <- unlist(strsplit(as.character(variants$FORMAT[i]), ":"))
  ws_info <- unlist(strsplit(as.character(variants$ws20171223_MPs_tomatoMuscle8wkQuiescent201[i]), ":"))

  # use the AD index to pull the pair from ws_info
  AD_pair <- ws_info[grep("AD", format_info)[1]]
  
  # split the pair by ,
  AD_split <- as.integer(unlist(strsplit(as.character(AD_pair), ",")))
  
  # then do the math x/(x+y) (this is a %)
  # store it in VAF
  VAF[i] <- AD_split[1] / (AD_split[1]+AD_split[2])
}

boxplot(VAF, main = "Distribution of VAFs Across Variants", ylab = "VAF")


VAF_greater5 <- which(VAF >= 0.05)

cat("How many variants have VAF > 5%?", length(VAF_greater5), "\n")

cat("How many of those variants (the ones with >5% VAF) are in coding regions?",length(VAF_greater5[grep("protein_coding", variants$INFO[VAF_greater5])]))

```
